backend/API_standards.md

```diff
@@ -0,0 +1,183 @@
+| PUT       | /questions/:id   | questions#update     | update a specific question                   |
+| DELETE    | /questions/:id   | questions#delete     | delete a specific question                   |
+
+Non CRUD operation will follow the same pattern, example:
+
+| HTTP Verb | Path              | Service#Function     | Used for                                     |
+| --------- | ----------------  | -------------------- | -------------------------------------------- |
+| GET       | /questions/search | questions#search     | search questions (filterable query string)   |
+
+Nested resources will follow the same pattern:
+
+| HTTP Verb | Path                 | Service#Function     | Used for                                     |
+| --------- | -------------------  | -------------------- | -------------------------------------------- |
+| GET       | /company/:id/members | members#index        | list all members for a specific company      |
+
+## API Responses
+
+All endpoints should return a seamless response, independent of the endpoint action or expected result values.
+
+### JSON object response
+Every endpoint should return a JSON object with the appropriate `payload key`, rather than  collection or plain values.
+
+Example:
+
+```json
+{
+  "questions": [{"id": 1, "description": "desc 1"},
+                {"id": 2, "description": "desc 2"}]
+}
+```
+
+### Pagination
+Pagination details would be sent as part of the same JSON object.
+Suggested query string inputs, `page_size` and `page_number`.
+
+Example `GET /questions?page_size=8&page_number=4`
+
+Response Example:
+```json
+{
+  "meta": {
+            "total_pages": 5,
+            "current_page": 1
+          },
+  "questions": [
+                 {"id": 1, "description": "desc 1"},
+                 {"id": 2, "description": "desc 2"}
+               ]
+}
+```
+
+### Error Handling
+
+All errors should follow the same structure, each error object in the collection represent one single error.
+
+The error object will always respond to the following attributes:
+- message
+- title
+
+But other (optional) attributes could be added when needed
+- attribute
+- position
+- code
+- url
+
+In general a `412 Precondition failed` will be used, please refer to the [Status Codes](status_codes) section for other scenarios references.
+
+Not Found example:
+```json
+{
+  "errors": [
+    {
+      "title": "Resource not found",
+      "message": "User not found, please check your identifier"
+    }
+  ]
+}
+```
+
+Validation error example:
+```json
+{
+  "errors": [
+    {
+      "title": "Invalid format",
+      "message": "The email haves an invalid format",
+      "attribute": "email"
+    },
+    {
+      "title": "Invalid format",
+      "message": "The password length should be at least 8 characters",
+      "attribute": "password"
+    }
+  ]
+}
+```
+
+For those fields accepting collections, the error should point to the specific index with errors, example:
+```json
+{
+  "errors": [
```

Comment:
>I have one question. I thought our error conventions will always be a map, Does it change?

Answering your question, yes, it would change from having 3-4 different error structures to just one: `{errors: [{:message :title :field etc...}]}`. 
Plus every error in the collection would have the same structure, so the API consumer doesn't have to parse a different JSON depending on the error type.